### ISSUE TITLE
C# decompilation: issues with complex conditions.

### ISSUE THREAD

**Issue by dvorm** [icsharpcode/ILSpy#1094]
The new C# decompiler from version 3.xx does a very poor job decompiling complex conditions. This makes longer methods very hard to read.

As a minimum, I'd like the option to switch back to the decompiler from version 2.4. Or, even better, fix the new decompiler. It has some nice features.

Here's an example from a game I'm modding.
Decompiled with version 2.4.0.1963:
```
		if ((((SHModel.Instance.Chapter == UnitChapter.DarkAngels || SHModel.Instance.Chapter == UnitChapter.DA2) && this.ClassID == UnitClassID.Apothecary) || ((SHModel.Instance.Chapter == UnitChapter.DarkAngels || SHModel.Instance.Chapter == UnitChapter.AngelsOfRedemption || SHModel.Instance.Chapter == UnitChapter.DA2) && this.ClassID == UnitClassID.Sergeant)) && base.Faction == UnitFaction.Faction1)
		{
			return 10;
		}
		return this.skillStats.GetIntValue(SkillDataStatsBase.PlasmaVolatileChanceBonus);

```
The same with version 3.1.0.3599-beta1

```
		if (SHModel.Instance.Chapter != UnitChapter.DarkAngels && SHModel.Instance.Chapter != UnitChapter.DA2)
		{
			goto IL_002d;
		}
		if (this.ClassID != UnitClassID.Apothecary)
		{
			goto IL_002d;
		}
		goto IL_006b;
		IL_007a:
		return this.skillStats.GetIntValue(SkillDataStatsBase.PlasmaVolatileChanceBonus);
		IL_006b:
		if (base.Faction == UnitFaction.Faction1)
		{
			return 10;
		}
		goto IL_007a;
		IL_002d:
		if ((SHModel.Instance.Chapter == UnitChapter.DarkAngels || SHModel.Instance.Chapter == UnitChapter.AngelsOfRedemption || SHModel.Instance.Chapter == UnitChapter.DA2) && this.ClassID == UnitClassID.Sergeant)
		{
			goto IL_006b;
		}
		goto IL_007a;

```

For completeness, here's the IL code:

```
	IL_0000: call class TXEngine.NetworkedCombatSystem.Model.SHModel TXEngine.NetworkedCombatSystem.Model.SHModel::get_Instance()
	IL_0005: callvirt instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitChapter TXEngine.NetworkedCombatSystem.Model.SHModel::get_Chapter()
	IL_000a: ldc.i4.5
	IL_000b: beq IL_0021

	IL_0010: call class TXEngine.NetworkedCombatSystem.Model.SHModel TXEngine.NetworkedCombatSystem.Model.SHModel::get_Instance()
	IL_0015: callvirt instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitChapter TXEngine.NetworkedCombatSystem.Model.SHModel::get_Chapter()
	IL_001a: ldc.i4.s 16
	IL_001c: bne.un IL_002d

	IL_0021: ldarg.0
	IL_0022: call instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitClassID TXEngine.NetworkedCombatSystem.Model.SHUnit::get_ClassID()
	IL_0027: ldc.i4.2
	IL_0028: beq IL_006b

	IL_002d: call class TXEngine.NetworkedCombatSystem.Model.SHModel TXEngine.NetworkedCombatSystem.Model.SHModel::get_Instance()
	IL_0032: callvirt instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitChapter TXEngine.NetworkedCombatSystem.Model.SHModel::get_Chapter()
	IL_0037: ldc.i4.5
	IL_0038: beq IL_005f

	IL_003d: call class TXEngine.NetworkedCombatSystem.Model.SHModel TXEngine.NetworkedCombatSystem.Model.SHModel::get_Instance()
	IL_0042: callvirt instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitChapter TXEngine.NetworkedCombatSystem.Model.SHModel::get_Chapter()
	IL_0047: ldc.i4.s 9
	IL_0049: beq IL_005f

	IL_004e: call class TXEngine.NetworkedCombatSystem.Model.SHModel TXEngine.NetworkedCombatSystem.Model.SHModel::get_Instance()
	IL_0053: callvirt instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitChapter TXEngine.NetworkedCombatSystem.Model.SHModel::get_Chapter()
	IL_0058: ldc.i4.s 16
	IL_005a: bne.un IL_007a

	IL_005f: ldarg.0
	IL_0060: call instance valuetype TXEngine.NetworkedCombatSystem.Model.UnitClassID TXEngine.NetworkedCombatSystem.Model.SHUnit::get_ClassID()
	IL_0065: ldc.i4.0
	IL_0066: bne.un IL_007a

	IL_006b: ldarg.0
	IL_006c: call instance valuetype ['TXEngine-NetworkedCombatSystem']TXEngine.NetworkedCombatSystem.UnitFaction ['TXEngine-NetworkedCombatSystem']TXEngine.NetworkedCombatSystem.GameUnit::get_Faction()
	IL_0071: ldc.i4.1
	IL_0072: bne.un IL_007a

	IL_0077: ldc.i4.s 10
	IL_0079: ret

	IL_007a: ldarg.0
	IL_007b: ldfld class ['Assembly-CSharp.dll']SkillDataStatsBase TXEngine.NetworkedCombatSystem.Model.SHUnit::skillStats
	IL_0080: ldsfld string SkillDataStatsBase::PlasmaVolatileChanceBonus
	IL_0085: callvirt instance int32 StatsHolder::GetIntValue(string)
	IL_008a: ret

```

**Comment by meirumeiru** [comment]
I see the same... and what I also see often is that "nothing" is decompiled into 10 lines of switch and if and else... and when you analyze the code generated, then you see that it's doing nothing...
things like
switch(3)
{
case 5:
goto blabla;
default:
break;
}
blabla:
... that wasn't like this in old versions. (I can give more information if needed)

**Comment by siegfriedpammer** [comment]
@meirumeiru 
> what I also see often is that "nothing" is decompiled into 10 lines of switch and if and else... and when you analyze the code generated, then you see that it's doing nothing...

Are you looking at obfuscated assemblies? It's a common obfuscation technique to add unreachable blocks/instructions that do nothing, just to confuse a decompiler. You should use a deobfuscator. ILSpy by definition does not do deobfuscation.

**Comment by siegfriedpammer** [comment]
I think this was fixed in #1176.
